src/bin/stratis-utils.rs

```diff
@@ -11,12 +11,85 @@ use std::{
     fmt::{self, Display},
 };
 
-use clap::{App, Arg};
+use clap::{App, AppSettings, Arg, SubCommand};
 use data_encoding::BASE32_NOPAD;
+use serde_json::{json, Value};
+
+use devicemapper::{Bytes, Sectors};
+
+use stratisd::engine::{crypt_metadata_size, ThinPoolSizeParams, BDA};
 
 #[cfg(feature = "systemd_compat")]
 use crate::generators::{stratis_clevis_setup_generator, stratis_setup_generator};
 
+const FS_SIZE_LOOKUP_TABLE_LEN: usize = 27;
+const FS_SIZE_START_POWER: usize = 29;
+const FS_LOGICAL_SIZE_MAX: u128 = 36_028_797_018_963_968; // 32 PiB
+const FS_LOGICAL_SIZE_MIN: u128 = 536_870_912; // 512 MiB
+
+struct FSSizeLookup {
+    internal: Vec<u128>,
+}
+
+impl FSSizeLookup {
+    fn lookup(&self, logical_size: Bytes) -> Sectors {
+        let raw_size = *logical_size;
+        assert!(raw_size >= FS_LOGICAL_SIZE_MIN);
+        assert!(raw_size < FS_LOGICAL_SIZE_MAX);
+        // At FS_LOGICAL_SIZE_MAX there is an 8 integer range of floating
+        // point value that the u128 value may occupy. Given the large size
+        // om the number and that it is the log that is being taken this is
+        // of no concern for the correctness of the value.
+        #[allow(clippy::cast_precision_loss)]
+        let lg = f64::log2(raw_size as f64);
```

Comment:
I seem to remember you discussed that the sizes in FSSizeLookup were for each doubling of the filesystem size. Given this information, log makes sense here but it may not be immediately evident to someone reading the code. I'd suggest adding documentation explaining this a bit.

Reply:
I've added a bunch of comments.